src/operations/blas1_trees.hpp

```diff
@@ -619,6 +619,278 @@ Rotg<operand_t>::adjust_access_displacement() {
+SYCL_BLAS_INLINE typename Rotmg<operand_t>::index_t Rotmg<operand_t>::get_size()
+    const {
+  return static_cast<Rotmg<operand_t>::index_t>(1);
+}
+
+/**
+ * For further details about the rotmg algorithm refer to:
+ *
+ * Hopkins, Tim (1998) Restructuring the BLAS Level 1 Routine for Computing the
+ * Modified Givens Transformation. Technical report. , Canterbury, Kent, UK.
+ *
+ * and
+ *
+ * C. L. Lawson, R. J. Hanson, D. R. Kincaid, and F. T. Krogh. Basic linear algebra
+ * subprograms for Fortran usage. ACM Trans. Math. Softw., 5:308-323, 1979.
+ */
+template<typename operand_t>
+SYCL_BLAS_INLINE typename Rotmg<operand_t>::value_t Rotmg<operand_t>::eval(
+    typename Rotmg<operand_t>::index_t i) {
+
+  using zero = constant<value_t, const_val::zero>;
+  using one = constant<value_t, const_val::one>;
+  using two = constant<value_t, const_val::two>;
+  using m_one = constant<value_t, const_val::m_one>;
+  using m_two = constant<value_t, const_val::m_two>;
+
+  using error = two;
+
+  using clts_matrix = one; /* co-sin less than sin */
+  using sltc_matrix = zero; /* sin less than co-sin */
+  using rescaled_matrix = m_one;
+  using unit_matrix = m_two;
+
+  /* Gamma is a magic number used to re-scale the output and avoid underflows or
+   * overflows. Consult the papers above for more info */
+  constexpr value_t gamma = static_cast<value_t>(4096.0);
+
+  /* Square root of gamma. It is hardcoded to avoid computing it on every call */
+  constexpr value_t gamma_sq = static_cast<value_t>(1.67772e7);
+
+  /* Inverse of the square root of gamma (i.e. 1 / sqrt(gamma)) */
+  constexpr value_t inv_gamma_sq = static_cast<value_t>(5.96046e-8);
+
+  value_t &d1_ref = d1_.eval(i);
+  value_t &d2_ref = d2_.eval(i);
+  value_t &x1_ref = x1_.eval(i);
+  value_t &flag_ref = param_.eval(static_cast<index_t>(0));
+  value_t &h11_ref = param_.eval(static_cast<index_t>(1));
+  value_t &h21_ref = param_.eval(static_cast<index_t>(2));
+  value_t &h12_ref = param_.eval(static_cast<index_t>(3));
+  value_t &h22_ref = param_.eval(static_cast<index_t>(4));
+
+  value_t d1 = d1_ref;
+  value_t d2 = d2_ref;
+  value_t x1 = x1_ref;
+  const value_t y1 = y1_.eval(i);
+
+  value_t flag;
+  value_t h11;
+  value_t h21;
+  value_t h12;
+  value_t h22;
+  value_t swap_temp;
+
+  /* d1 cannot be negative.
+   * Note: negative d2 is valid as a way to implement row removal */
+  if (d1 < zero::value()) {
+    flag = error::value();
+  }
+  /* If the input is of the form (c, 0), then we already have the expected output.
+   * No calculations needed in this case */
+  else if (d2 == zero::value() || y1 == zero::value()) {
+    flag = unit_matrix::value();
+  }
+  /* If the input is of the form (0, c) - just need to swap the elements.
+   * Scaling may be needed */
+  else if ((d1 == zero::value() || x1 == zero::value()) && d2 > zero::value()) {
+    flag = clts_matrix::value();
+    /* clts_matrix assumes h12 and h21 values. But they still need to be set
```

Comment:
I'm not seeing where this happens. Can you say it will happen later and label it perhaps?

Reply:
If you go to the scaling part of the algorithm. The matrix will be rescaled based on the value of d1 or d2. In those cases. h12 or h21 might be rescaled as well. Where would you like me to put the label?

Reply:
As discussed on private chate, updated the comment on the rescale section to be more clear